CSharp/Library/Microsoft.Bot.Builder/Dialogs/PromptRecognizers.cs

```diff
@@ -0,0 +1,468 @@
+        public IDictionary<string, T> Locales { get; set; }
+
+        public LocalizedCache()
+        {
+            this.Locales = new ConcurrentDictionary<string, T>();
+        }
+        public LocalizedCache(string key)
+            : this()
+        {
+            this.Id = key;
+        }
+    }
+
+    public class LocalizedCacheList<T> : List<LocalizedCache<T>>
+    {
+        public LocalizedCache<T> this[string key]
+        {
+            get { return this.SingleOrDefault(x => x.Id == key); }
+        }
+    }
+
+    public class ChronoDuration
+    {
+        public string Entity { get; internal set; }
+        public ChronoDurationResolution Resolution { get; set; }
+
+        public ChronoDuration()
+        {
+            this.Resolution = new ChronoDurationResolution();
+        }
+    }
+
+    public class ChronoDurationResolution
+    {
+        public string ResolutionType { get; set; }
+        public DateTime? Start { get; set; }
+        public DateTime? End { get; set; }
+    }
+
+    public interface IPromptRecognizers
+    {
+        IEnumerable<RecognizeEntity<string>> RecognizeLocalizedRegExp(IMessageActivity context, string expressionKey, ResourceManager resourceManager);
+
+        IEnumerable<RecognizeEntity<T>> RecognizeChoices<T>(string utterance, IDictionary<T, IEnumerable<T>> choices, IPromptRecognizeChoicesOptions options = null);
+
+        IEnumerable<RecognizeEntity<string>> RecognizeLocalizedChoices(IMessageActivity context, string choicesKey, ResourceManager resourceManager, IPromptRecognizeChoicesOptions options = null);
+
+        IEnumerable<RecognizeEntity<double>> RecognizeNumbers(IMessageActivity context, IPromptRecognizeNumbersOptions options = null);
+
+        IEnumerable<RecognizeEntity<long>> RecognizeOrdinals(IMessageActivity context);
+
+        IEnumerable<RecognizeEntity<string>> RecognizeTimes(IMessageActivity context);
+        
+        IEnumerable<RecognizeEntity<bool>> RecognizeBooleans(IMessageActivity context);
+    }
+
+    [Serializable]
+    public class DefaultRecognizers : IPromptRecognizers
+    {
+        private const string resource_key_cardinals = "NumberTerms";
+        private const string resource_key_ordinals = "NumberOrdinals";
+        private const string resource_key_reverser_ordinals = "NumberReverseOrdinals";
+        private const string resource_key_number_regex = "NumberExpression";
+        private const string resource_key_booleans = "BooleanChoices";
+
+        private static Regex simpleTokenizer = new Regex(@"\w+", RegexOptions.IgnoreCase);
+        private static IDictionary<string, IDictionary<string, Regex>> expCache = new ConcurrentDictionary<string, IDictionary<string, Regex>>();
+        private static IDictionary<string, IDictionary<string, IDictionary<string, IEnumerable<string>>>> choicesCache = new ConcurrentDictionary<string, IDictionary<string, IDictionary<string, IEnumerable<string>>>>();
+
+        public DefaultRecognizers()
```

Comment:
it might be good to have those helper functions as extensions methods. 

---
In reply to: [112468611](https://github.com/Microsoft/BotBuilder/pull/2609#discussion_r112468611) [](ancestors = 112468611)